File: apps/scala/src/main/scala/machinelearning/chapter10/RandomForest.scala
package machinelearning.chapter10

import java.util.Random
import machinelearning.chapter02.Features
import machinelearning.chapter03.DecisionTree

/** 学習した 1 本の木と、その木が使った列・行。 */
case class FittedTree(columns: Vector[String], rows: Vector[Int], model: DecisionTree)

/** 第 3 章の決定木をブートストラップ標本と特徴量の部分集合で学習し、多数決で予測するランダムフォレスト。 */
class RandomForest(nEstimators: Int, maxFeatures: Int, maxDepth: Option[Int], seed: Long)
    extends Classifier:
  private var fittedTrees: Vector[FittedTree] = Vector.empty

  /** 学習した決定木。 */
  def trees: Vector[FittedTree] = fittedTrees

  override def fit(x: Vector[Features], t: Vector[String]): RandomForest =
    val random = Random(seed)
    val allColumns = x.head.columns
    fittedTrees = (0 until nEstimators).toVector.map { _ =>
      val rows = RandomForest.bootstrapSample(x.size, random)
      // Java 版（Collections.shuffle）と同じ手順で列を選ぶ
      val shuffled = RandomForest.shuffle(allColumns, random)
      val chosen = shuffled.take(maxFeatures).toSet
      val columns = allColumns.filter(chosen.contains)
      val sampleX = rows.map(row => RandomForest.selectColumns(x(row), columns))
      val sampleT = rows.map(t)
      val tree = maxDepth.fold(DecisionTree.unlimited())(DecisionTree.withMaxDepth)
      FittedTree(columns, rows, tree.fit(sampleX, sampleT))
    }
    this

  override def predict(x: Vector[Features]): Vector[String] =
    if fittedTrees.isEmpty then throw IllegalStateException("fit で学習してから predict を呼んでください")
    RandomForest.majorityVote(
      fittedTrees.map(tree => tree.model.predict(RandomForest.selectColumns(x, tree.columns)))
    )

object RandomForest:
  /** 深さを制限しない決定木の森。 */
  def of(nEstimators: Int, maxFeatures: Int, seed: Long): RandomForest =
    RandomForest(nEstimators, maxFeatures, None, seed)

  /** 深さの上限を指定した決定木の森。 */
  def withMaxDepth(nEstimators: Int, maxFeatures: Int, maxDepth: Int, seed: Long): RandomForest =
    require(maxDepth >= 0, "深さの上限は 0 以上にしてください")
    RandomForest(nEstimators, maxFeatures, Some(maxDepth), seed)

  /** サンプルごとに、最も多い予測を選ぶ。同数なら先に現れた予測を選ぶ。 */
  def majorityVote(votes: Vector[Vector[String]]): Vector[String] =
    votes.head.indices.toVector.map(sample => mostCommon(votes.map(_(sample))))

  /** 0 から size - 1 までの行番号を、重複を許して size 個選ぶ。 */
  def bootstrapSample(size: Int, random: Random): Vector[Int] =
    Vector.fill(size)(random.nextInt(size))

  /** 特徴量から、指定した列だけを取り出す。 */
  def selectColumns(features: Features, columns: Vector[String]): Features =
    Features(columns, columns.map(features.value))

  def selectColumns(x: Vector[Features], columns: Vector[String]): Vector[Features] =
    x.map(selectColumns(_, columns))

  /** Java の Collections.shuffle と同じ手順で並べ替える。 */
  private[chapter10] def shuffle[A](items: Vector[A], random: Random): Vector[A] =
    val array = items.toArray[Any]
    for i <- array.length - 1 to 1 by -1 do
      val j = random.nextInt(i + 1)
      val tmp = array(i)
      array(i) = array(j)
      array(j) = tmp
    array.toVector.map(_.asInstanceOf[A])

  private def mostCommon(labels: Vector[String]): String =
    val counts = labels.foldLeft(scala.collection.immutable.ListMap.empty[String, Int]) {
      (acc, label) => acc.updated(label, acc.getOrElse(label, 0) + 1)
    }
    counts.maxBy(_._2)._1

File: apps/php/src/Chapter10/RandomForest.php
<?php

declare(strict_types=1);

namespace GettingStartedMl\Chapter10;

use GettingStartedMl\Chapter03\DecisionTree;
use GettingStartedMl\Chapter10;
use GettingStartedMl\Random;

/**
 * 第 3 章の決定木を、ブートストラップ標本と特徴量の部分集合で何本も育てた森。
 *
 * 予測は木ごとの多数決。Rubix ML にもランダムフォレストがあるので、
 * Elixir 版と違って「自作しかない」章にはならない。
 */
final readonly class RandomForest implements Predictor
{
    /** @param list<ForestTree> $trees */
    public function __construct(public array $trees)
    {
    }

    /**
     * 木を n_estimators 本学習する。maxDepth が null なら深さの上限なし。
     *
     * @param list<array<string, float>> $x
     * @param list<string>               $t
     * @param list<string>               $columns
     */
    public static function fit(
        array $x,
        array $t,
        array $columns,
        int $nEstimators,
        int $maxFeatures,
        ?int $maxDepth,
        int $seed,
    ): self {
        $random = new Random($seed);
        $trees = [];

        for ($i = 0; $i < $nEstimators; ++$i) {
            $rows = Chapter10::bootstrapSample(count($x), $random);
            $chosen = array_slice(self::shuffleWith($columns, $random), 0, $maxFeatures);
            // 列の順は元のまま残す。
            $treeColumns = array_values(
                array_filter($columns, static fn (string $c): bool => in_array($c, $chosen, true)),
            );

            $sampleX = [];
            $sampleT = [];

            foreach ($rows as $row) {
                $sampleX[] = Chapter10::takeColumns($x[$row], $treeColumns);
                $sampleT[] = $t[$row];
            }

            $trees[] = new ForestTree(
                $treeColumns,
                $rows,
                DecisionTree::fit($sampleX, $sampleT, $treeColumns, $maxDepth),
            );
        }

        return new self($trees);
    }

    /**
     * 木ごとの予測を多数決でまとめる。
     *
     * @param list<array<string, float>> $x
     *
     * @return list<string>
     */
    public function predict(array $x): array
    {
        $votes = array_map(
            static fn (ForestTree $tree): array => $tree->tree->predict(
                array_map(
                    static fn (array $features): array => Chapter10::takeColumns($features, $tree->columns),
                    $x,
                ),
            ),
            $this->trees,
        );

        return Chapter10::majorityVote($votes);
    }

    /**
     * 渡した乱数生成器で Fisher-Yates の並べ替えをする。
     *
     * 第 2 章の Random::shuffle はシードから始めるが、森を作る途中では
     * 状態を引き継ぐ必要があるので、生成器そのものを受け取る。
     *
     * @param list<string> $items
     *
     * @return list<string>
     */
    private static function shuffleWith(array $items, Random $random): array
    {
        for ($i = count($items) - 1; $i >= 1; --$i) {
            $j = $random->nextInt($i + 1);
            [$items[$i], $items[$j]] = [$items[$j], $items[$i]];
        }

        return array_values($items);
    }
}
